lib/propolis/src/hw/virtio/p9fs.rs

```diff
@@ -0,0 +1,986 @@
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+
+use crate::common::*;
+use crate::hw::pci;
+use crate::migrate::Migrator;
+use crate::util::regmap::RegMap;
+use crate::vmm::MemCtx;
+
+use super::bits::*;
+use super::pci::{PciVirtio, PciVirtioState};
+use super::queue::{write_buf, Chain, VirtQueue, VirtQueues};
+use super::VirtioDevice;
+
+use ispf::WireSize;
+use lazy_static::lazy_static;
+use libc::{
+    DT_DIR, DT_REG, EILSEQ, EINVAL, ENOENT, ENOLCK, ENOTSUP, EOVERFLOW, ERANGE,
+};
+use num_enum::TryFromPrimitive;
+use p9ds::proto::{
+    self, Dirent, MessageType, P9Version, Qid, QidType, Rattach, Rclunk,
+    Rgetattr, Rlerror, Rlopen, Rread, Rreaddir, Rstatfs, Rwalk, Tattach,
+    Tgetattr, Tlopen, Tread, Treaddir, Tstatfs, Twalk, Version,
+    P9_GETATTR_BASIC,
+};
+use slog::{warn, Logger};
+
+#[usdt::provider(provider = "propolis")]
+mod probes {
+    fn p9fs_cfg_read() {}
+}
+
+/// This is a work-in-progres P9 filesystem device. It's a minimum viable
+/// implementation provide a P9 filesystem to guest. It's been tested with
+/// illumos and Linux guests. There are many capabilities that are not yet
+/// implemented.
+///
+/// The design centers around a P9Handler trait that allows various different
+/// types of P9 devices to be implemented. This file includes a `HostFSHandler`
+/// implementation that allows mounting host filesystems in the guest.
+/// Currently filesystems can only be mounted as read-only. Another
+/// implementation is in the SoftNpu device that supports P4 program transfer
+/// via p9fs.
+pub struct PciVirtio9pfs {
+    virtio_state: PciVirtioState,
+    pci_state: pci::DeviceState,
+    handler: Arc<dyn P9Handler>,
+}
+
+impl PciVirtio9pfs {
+    pub fn new(queue_size: u16, handler: Arc<dyn P9Handler>) -> Arc<Self> {
+        let queues = VirtQueues::new(
+            NonZeroU16::new(queue_size).unwrap(),
+            NonZeroU16::new(1).unwrap(),
+        );
+        let msix_count = Some(2); //guess
+        let (virtio_state, pci_state) = PciVirtioState::create(
+            queues,
+            msix_count,
+            VIRTIO_DEV_9P,
+            VIRTIO_SUB_DEV_9P_TRANSPORT,
+            pci::bits::CLASS_STORAGE,
+            VIRTIO_9P_CFG_SIZE,
+        );
+        Arc::new(Self { virtio_state, pci_state, handler })
+    }
+}
+
+impl VirtioDevice for PciVirtio9pfs {
+    fn cfg_rw(&self, mut rwo: RWOp) {
+        P9FS_DEV_REGS.process(&mut rwo, |id, rwo| match rwo {
+            RWOp::Read(ro) => {
+                probes::p9fs_cfg_read!(|| ());
+                match id {
+                    P9fsReg::TagLen => {
+                        ro.write_u16(self.handler.target().len() as u16);
+                    }
+                    P9fsReg::Tag => {
+                        let mut bs = [0; 256];
+                        for (i, x) in self.handler.target().bytes().enumerate()
+                        {
+                            bs[i] = x as u8;
+                        }
```

Comment:
I should have caught this earlier: this could probably be reduced to a `as_bytes()` call on the `&str`, rather than a buffer copying situation.  In both cases, some bounds checking is likely needed to avoid potential overflow.

Reply:
When I change this to 
```
P9fsReg::Tag => {
    ro.write_bytes(self.handler.target().as_bytes());
    ro.fill(0);
}
```

The result is
```
thread 'vcpu-0' panicked at 'source slice length (14) does not match destination slice length (256)', lib/propolis/src/common.rs:156:29
stack backtrace:
   0: rust_begin_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
   2: core::slice::<impl [T]>::copy_from_slice::len_mismatch_fail
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/slice/mod.rs:3239:13
   3: propolis::common::ReadOp::write_bytes
   4: propolis::util::regmap::RegMap<ID>::process
   5: <propolis::hw::virtio::p9fs::PciVirtio9pfs as propolis::hw::virtio::VirtioDevice>::cfg_rw
   6: propolis::util::regmap::RegMap<ID>::process
   7: propolis::hw::pci::device::<impl propolis::hw::pci::Endpoint for D>::bar_rw
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I've added a check to the current code to avoid overruning the buffer for tag strings longer than 256 bytes.